app/src/test/java/com/willowtree/vocable/utils/FaceTrackingPermissionsTest.kt

```diff
@@ -66,60 +61,155 @@ class FaceTrackingPermissionsTest {
+        val permissions = createFaceTrackingPermissions(
+            shouldShowRequestPermissionRationale = true
+        )
         advanceUntilIdle()
 
         permissions.requestFaceTracking()
-        assertEquals(permissions.permissionState.first(), IFaceTrackingPermissions.PermissionState.PermissionRequested)
 
-        fakePermissionRegisterForLaunch.triggerActivityResult(contract = Manifest.permission.CAMERA, result = true)
-        assertEquals(permissions.permissionState.first(), IFaceTrackingPermissions.PermissionState.Enabled)
+        assertEquals(1, fakePermissionsDialogShower.showPermissionRationaleDialogCalledCount)
+        assertEquals(0, fakePermissionsDialogShower.showSettingsPermissionDialogCalledCount)
     }
 
     @Test
-    fun `when requestFaceTracking unsuccessful, permission state set to disabled`() = runTest {
-        val sharedPreference = createSharedPrefs(headTrackingEnabled = false)
-        val permissions = createFaceTrackingPermissions(sharedPreference)
+    fun `when rationale should be shown, requestFaceTracking() shows dialog and onPositiveClick launches permissionLauncher`() =
+        runTest {
+            val permissions = createFaceTrackingPermissions(
+                shouldShowRequestPermissionRationale = true
+            )
+            advanceUntilIdle()
+
+            permissions.requestFaceTracking()
+
+            assertEquals(1, fakePermissionsDialogShower.showPermissionRationaleDialogCalledCount)
+            assertEquals(0, fakePermissionsDialogShower.showSettingsPermissionDialogCalledCount)
+            fakePermissionsDialogShower.permissionDialogOnPositiveClick.invoke()
+            assertEquals(1, fakePermissionRegisterForLaunch.launchCount)
+        }
+
+    @Test
+    fun `when permission rationale should be shown, rationale dialog shown and onNegativeClick launches permissionLauncher`() =
+        runTest {
+            val sharedPreferences = createSharedPrefs(headTrackingEnabled = true)
+
+            // requestFaceTracking() called implicitly in init block when headTrackingEnabled is true
+            createFaceTrackingPermissions(
+                sharedPreferences = sharedPreferences,
+                shouldShowRequestPermissionRationale = true
+            )
+            advanceUntilIdle()
+
+            assertEquals(1, fakePermissionsDialogShower.showPermissionRationaleDialogCalledCount)
+            assertEquals(0, fakePermissionsDialogShower.showSettingsPermissionDialogCalledCount)
+            fakePermissionsDialogShower.permissionDialogOnNegativeClick.invoke()
+            assertFalse(
+                "Head tracking should be disabled if user denies dialog",
+                sharedPreferences.getHeadTrackingEnabled()
+            )
+        }
+
+    @Test
+    fun `given initial rationale dialog denied, settings dialog rationale shown and updates head tracking when enabled`() = runTest {
```

Comment:
This is a magilla of a test, but at the same time, I don't see a whole lot of other options for testing this fiddly permission stuff, and I like having these flows tested so 🤷

Reply:
I split out this test into two so it can at least verify head tracking permissions vs dialogs being shown separately. still doing a lot because of "fiddly permission stuff" reasons but should make it easier to track down reasons for test failure